### PR TITLE
Blacklist support

### PR DESCRIPTION
**Pull request by jrs65**
For the moment I have (begrudgingly) gone for the non RESTful `/blacklist` and `/update-blacklist` endpoints as I figured we could add that support in a later pull request.

### CODE REVIEW

**Comment by jrs65** [comment]
Force pushed some changes that move the unit tests to the right location.

File: coco/worker.py [hunk 1]
```diff
@@ -74,41 +71,74 @@ def main_loop(endpoints, forwarder, coco_port, metrics_port, log_level):
...
+                code = 200
+
+            # Process a known exception source into a response
+            except CocoException as e:
+                result = e.to_dict()
+                code = e.status_code
+
+            # Unexpected exceptions are returned as HTTP 500 errors, and dump a
+            # traceback
+            except Exception as e:
```
Comment (nritsche) [inline coco/worker.py]:
```suggestion
            except BaseException as e:
```

to include SystemExit, KeyboardInterrupt, GeneratorExit (https://docs.python.org/3/library/exceptions.html#exception-hierarchy)?

**Comment by nritsche** [comment]
As discussed, the blacklist should be ignored when hosts are explicitly specified to an endpoint call. But we can add that later.

File: .travis.yml [hunk 2]
```diff
@@ -38,7 +38,7 @@ script:
     - black -l 99 --check --exclude kotekan .
 
     # Run pydocstyle on all .py files
-    - find . -type d \( -path ./docs -o -path ./tests/simulate-chime \) -prune -o ! -name versioneer.py ! -name _version.py -name \*.py -exec pydocstyle --convention=numpy {} +
+    - find . -type d \( -path ./docs -o -path ./tests/simulate-chime \) -prune -o ! -name versioneer.py ! -name _version.py -name \*.py -exec pydocstyle --convention=numpy --add-ignore=D105,D202 {} +
```
[the remark this answers is not in the document]
Comment (jrs65) [reply .travis.yml]: Change it back if you like, but you have to fix the errors! I consider those ones particularly stupid:
- Forbidding any newlines between the docstring and code in a function.
- Requiring documentation for dunder methods like hash and eq.

File: coco/blacklist.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        success
+            Did the update succeed.
+        """
+
+        h, checks = self._check_hosts(hosts)
+
+        if not all(checks):
+            bad_hosts = [host for host, check in zip(hosts, checks) if not check]
+            msg = f"Could not add to blacklist. Requested hosts {bad_hosts} unknown."
+            logger.info(msg)  # TODO: should this be logged here?
```
[the remark this answers is not in the document]
Comment (jrs65) [reply coco/blacklist.py]: I think the question is whether you want to log what happened at the server end in addition to what the client did. I would say yes, but maybe there's a better way to be doing that.

File: coco/blacklist.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+            raise InvalidUsage(
+                "Could not remove from blacklist as some hosts unknown.", context=bad_hosts
+            )
+
+        hosts = set(h)
+
+        not_blacklisted = hosts - self.hosts
+        if not_blacklisted:
+            logger.debug(f"Hosts {Host.print_list(not_blacklisted)} are not in " "the blacklist.")
```
[the remark this answers is not in the document]
Comment (jrs65) [reply coco/blacklist.py]: Ah, yeah, I guess `black` decided to re-merge the line.

File: coco/worker.py (hunk 1 above)
Comment (jrs65) [reply coco/worker.py]: Don't we want those exceptions to cause a genuine exit from the mainloop?

File: tests/test_blacklist.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+"""Test the blacklist, in case you didn't get that from the filename."""
```
[the remark this answers is not in the document]
Comment (jrs65) [reply tests/test_blacklist.py]: I would say yes.

File: .travis.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (nritsche) [reply .travis.yml]: I kind of agree with the magic methods.

I will change this:
"There's no blank line either before or after the docstring." ([PEP257](https://www.python.org/dev/peps/pep-0257/))

File: tests/test_blacklist.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nritsche) [reply tests/test_blacklist.py]: done

File: coco/worker.py (hunk 1 above)
Comment (nritsche) [reply coco/worker.py]: Ah, yes. I was just thinking the client should get a reply instead of timing out. We could re-raise in that case.

Comment (jrs65) [reply coco/worker.py]: Fair enough. I can add a final clause that catches, replies, and re-raises any base exception. I think I need to remove the reply into a finally block

**Comment by nritsche** [comment]
I don't understand why this happened now, but `black` wants to add this black line after the docstring that `pydocstyle` complains about :confused: 

I guess it means you win, we can ignore `D202` :smile:

File: coco/blacklist.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrs65) [reply coco/blacklist.py]: In any case I'll change this to debug.

**Pull request merged**